presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -311,47 +406,160 @@ private void processPage(Page page)
+                if (isSpillingEnabled()) {
+                    systemMemoryContext.setBytes(memorySize);
+
+                    if (shouldSpill(memorySize)) {
+                        log.info(
+                                "Started spilling memory %s",
+                                DataSize.succinctBytes(memorySize));
+
+                        spillToDisk();
+                    }
+                }
+                else {
+                    operatorContext.setMemoryReservation(memorySize);
+                }
                 return false;
             }
         }
 
-        public Iterator<Page> build()
+        private boolean shouldSpill(long memorySize)
+        {
+            return (maxEntriesBeforeSpill > 0 && groupByHash.getGroupCount() > maxEntriesBeforeSpill)
+                    || (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+        }
+
+        private boolean isSpillingEnabled()
+        {
+            return maxEntriesBeforeSpill > 0 || memorySizeBeforeSpill > 0;
+        }
+
+        public Iterator<Page> buildResult()
+        {
+            checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+            if (!spiller.isPresent()) {
+                return buildResultFromMemory();
+            }
+
+            try {
+                // TODO: don't spill here to disk, instead merge disk content with memory content
+                spillToDisk().get();
+                return mergeFromDisk();
+            }
+            catch (InterruptedException | ExecutionException e) {
+                Thread.currentThread().interrupt();
+                throw Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            closeSpiller();
+        }
+
+        private CompletableFuture<?> spillToDisk()
         {
-            List<Type> types = new ArrayList<>(groupByHash.getTypes());
             for (Aggregator aggregator : aggregators) {
-                types.add(aggregator.getType());
+                aggregator.setOutputPartial();
             }
 
+            if (!spiller.isPresent()) {
+                spiller = Optional.of(spillerFactory.get().create(buildTypes()));
+            }
+
+            // we are starting spilling process with current content of the aggregators and groupByHash...
+            busy = spiller.get().spill(buildResultFromMemory(sortedGroupIds(), buildTypes(), groupByHash, aggregators));
+            // ... and we immediately create new aggregators and groupByHash so we effectively transfer ownership
+            // over groupByHash and aggregators from this thread to spilling thread
+            rebuildAggregators();
+
+            return busy;
+        }
+
+        private Iterator<Page> mergeFromDisk()
+        {
+            checkState(spiller.isPresent());
+
+            convertToMerge();
+            rebuildAggregatorsForMerge();
```

Comment:
Indeed first call in line 520 is redundant, but it was simpler to implement it. This call here is to make sure that state of our class after converting to merge is consistent (aggregators are in the same state as the operator).